File: python/08_026.py
import heapq

def final_cache_state(n, operations, delays):
   # Initialize caches for each node
   caches = [{} for _ in range(n)]
   
   # Priority queue to simulate events with timestamps
   event_queue = []
   
   # Add initial operations to queue
   for node_id, key, value, timestamp in operations:
      # Process operation at source node immediately
      heapq.heappush(event_queue, 
                    (timestamp, node_id, node_id, key, 
                     value, timestamp))
      
      # Schedule propagation to other nodes
      for target_node in range(n):
         if target_node != node_id:
            arrival_time = timestamp + delays[node_id][target_node]
            heapq.heappush(event_queue, 
                          (arrival_time, target_node, node_id, key, 
                           value, timestamp))
   
   # Process all events
   while event_queue:
      arrival_time, target_node, source_node, key, value, op_timestamp = heapq.heappop(event_queue)
      
      # Apply the operation if it's newer than what we have
      if (key not in caches[target_node] or 
          caches[target_node].get(key + "_timestamp", 0) < op_timestamp):
         caches[target_node][key] = value
         caches[target_node][key + "_timestamp"] = op_timestamp
   
   # Clean up internal timestamp tracking
   for cache in caches:
      for key in list(cache.keys()):
         if key.endswith("_timestamp"):
            del cache[key]
   
   return caches

# Test with example
nodes = 3
operations = [
   (0, "x", 10, 1),  # Node 0 sets x=10 at time 1
   (1, "y", 20, 2),  # Node 1 sets y=20 at time 2
   (2, "x", 30, 3),  # Node 2 sets x=30 at time 3
   (0, "y", 40, 4)   # Node 0 sets y=40 at time 4
]
delays = [
   [0, 2, 3],  # Delays from node 0 to others
   [2, 0, 1],  # Delays from node 1 to others
   [3, 1, 0]   # Delays from node 2 to others
]

print(final_cache_state(nodes, operations, delays))

File: data structures and algorithms/cs/08_026.cs
using System;
using System.Collections.Generic;

class Operation {
   public int ArrivalTime { get; set; }
   public int TargetNode { get; set; }
   public int SourceNode { get; set; }
   public string Key { get; set; }
   public int Value { get; set; }
   public int OpTimestamp { get; set; }

   public Operation(
      int arrivalTime, 
      int targetNode, 
      int sourceNode, 
      string key, 
      int value, 
      int opTimestamp
   ) {
      ArrivalTime = arrivalTime;
      TargetNode = targetNode;
      SourceNode = sourceNode;
      Key = key;
      Value = value;
      OpTimestamp = opTimestamp;
   }
}

public class DistributedCacheConsistency {
   public static List<Dictionary<string, int>> 
   FinalCacheState(
      int n, 
      int[][] operationsData, 
      int[][] delays
   ) {
      var operations = new List<Operation>();
      foreach (var op in operationsData) {
         int nodeId = op[0];
         string key = "key" + op[1];
         int value = op[2];
         int timestamp = op[3];
         operations.Add(
            new Operation(
               0, nodeId, nodeId, key, value, timestamp
            )
         );
      }

      var caches = new List<Dictionary<string, int>>();
      var timestamps = new List<Dictionary<string, int>>();
      for (int i = 0; i < n; i++) {
         caches.Add(new Dictionary<string, int>());
         timestamps.Add(new Dictionary<string, int>());
      }

      var eventQueue = new PriorityQueue<Operation>(
         (op1, op2) => op1.ArrivalTime.CompareTo(op2.ArrivalTime)
      );

      foreach (var op in operations) {
         int nodeId = op.TargetNode;
         eventQueue.Enqueue(op);

         for (int targetNode = 0; targetNode < n; targetNode++) {
            if (targetNode != nodeId) {
               int arrivalTime = op.OpTimestamp + delays[nodeId][targetNode];
               eventQueue.Enqueue(
                  new Operation(
                     arrivalTime, targetNode, nodeId, op.Key, op.Value, op.OpTimestamp
                  )
               );
            }
         }
      }

      while (eventQueue.Count > 0) {
         var op = eventQueue.Dequeue();
         int targetNode = op.TargetNode;
         string key = op.Key;
         int value = op.Value;
         int opTimestamp = op.OpTimestamp;

         if (!timestamps[targetNode].ContainsKey(key) || timestamps[targetNode][key] < opTimestamp) {
            caches[targetNode][key] = value;
            timestamps[targetNode][key] = opTimestamp;
         }
      }

      return caches;
   }

   public static void Main(string[] args) {
      int nodes = 3;
      int[][] operations = {
         new int[] { 0, 0, 10, 1 },
         new int[] { 1, 1, 20, 2 },
         new int[] { 2, 0, 30, 3 },
         new int[] { 0, 1, 40, 4 }
      };
      int[][] delays = {
         new int[] { 0, 2, 3 },
         new int[] { 2, 0, 1 },
         new int[] { 3, 1, 0 }
      };

      var result = FinalCacheState(nodes, operations, delays);

      for (int i = 0; i < result.Count; i++) {
         Console.WriteLine("Node " + i + ": " + string.Join(", ", result[i]));
      }
   }
}